--- a/arch/mips/Makefile	2019-01-29 10:21:09.000000000 +0800
+++ b/arch/mips/Makefile	2019-04-29 22:26:52.760561714 +0800
@@ -185,7 +185,7 @@
 			-Wa,-mips64r2 -Wa,--trap
 
 ifdef CONFIG_2K1000_FIXUP_ALL
-cflags-$(CONFIG_CPU_LOONGSON2K)	+= -Wa$(comma)-mfix-loongson2k-jump=2
+//cflags-$(CONFIG_CPU_LOONGSON2K)	+= -Wa$(comma)-mfix-loongson2k-jump=2
 endif
 
 ifeq (,$(findstring march=octeon, $(cflags-$(CONFIG_CPU_CAVIUM_OCTEON))))
